src/Substance.hs

```diff
@@ -4,535 +4,450 @@
+
+subStmt, decl, bind, applyP :: Parser SubStmt
+subStmt = try bind <|> try decl <|> applyP
+decl = do
+  t' <- tParser
+  v' <- varParser
+  return (Decl t' v')
+bind = do
+  v' <- varParser
+  eq
+  e' <- exprParser
+  return (Bind v' e')
+applyP = do
+  p <- predicateParser
+  return (ApplyP p)
+
+----------------------------------------- Substance Typechecker ---------------------------
 
 -- | 'check' is the top-level semantic checking function. It takes a Substance
--- program as the input, checks the validity of the program, and outputs
--- a collection of information.
--- The check is done in two passes. First check all the declarations of
--- stand-alone objects like `Set`, and then check for all other things
-check :: SubProg -> SubEnv
-check p = let env1 = foldl checkDecls initE p
-              env2 = foldl checkReferencess env1 p
-              in
-            env2 { subObjs = reverse $ subObjs env2 }
-          where initE = SubEnv { subObjs = [], subDefs = M.empty,  subSymbols = M.empty, subApps = [], subArgs = M.empty }
-
-applyDef (n, m) d = case M.lookup n d of
+-- AST as the input and the environment recived from the DSLL type checker, checks the validity of the program acoording to
+-- the typechecking rules, and outputs a collection of information.
+check :: SubProg -> VarEnv -> VarEnv
+check p varEnv = let env = foldl checkSubStmt varEnv p
+                 in if (null (errors env))
+                    then env
+                    else error ("Substance type checking failed with the following problems: \n" ++ (errors env))
+
+checkSubStmt :: VarEnv -> SubStmt -> VarEnv
+checkSubStmt varEnv  (Decl t (VarConst n)) = let env  = checkT varEnv t
+                                                 env1 = addDeclaredName n env
+                                             in env1 { varMap = M.insert (VarConst n) t $ varMap env1 }
+
+checkSubStmt varEnv  (Bind v e) = let (vstr, vt) = checkVarE varEnv v
+                                      (estr, et) = checkExpression varEnv e -- TODO: Check lazy evaluation on et
+                                  in if (isJust vt && isJust et && vt /= et)
+                                     then varEnv { errors = (errors varEnv) ++ vstr ++ estr ++ "Expression of type " 
+                                                   ++ (show et)
+                                                   ++ " assigned to variable of type " ++ (show vt) ++ "\n"}
+                                     else varEnv { errors = (errors varEnv) ++ vstr ++ estr }
+
+checkSubStmt varEnv  (ApplyP p) = checkPredicate varEnv p
+
+checkPredicate :: VarEnv -> Predicate -> VarEnv
+checkPredicate varEnv (Predicate (PredicateConst p) args pos) =
+               case checkAndGet p (predicates varEnv) pos of
+               Right p -> case p of
+                 Pred1 p1 -> checkVarPred varEnv args p1
+                 Pred2 p2 -> checkRecursePred varEnv args
+               Left _ -> case checkAndGet p (operations varEnv) pos of
```

Comment:
I don't understand why checkPred ignores the first error and then calls checkAndGet again with *operations*. preds and ops shouldn't be mixed, right?